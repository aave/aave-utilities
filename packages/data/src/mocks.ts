import BigNumber from 'bignumber.js';
import {
  ReservesIncentiveDataHumanized,
  UserReservesIncentivesDataHumanized,
} from './formatters/incentive';
import { ReserveData } from './formatters/reserve';
import { UserReserveData } from './formatters/user';
import { RAY } from './ray.math';

export class ReserveMock {
  public reserve: ReserveData;

  private readonly config: { decimals: number };

  constructor(config: { decimals: number } = { decimals: 18 }) {
    this.config = config;
    this.reserve = {
      decimals: config.decimals,
      reserveFactor: '0',
      baseLTVasCollateral: '5000', // 50%
      averageStableRate: '0',
      stableDebtLastUpdateTimestamp: 1,
      liquidityIndex: RAY.toString(),
      reserveLiquidationThreshold: '6000', // 60%
      reserveLiquidationBonus: '0',
      variableBorrowIndex: RAY.toString(),
      variableBorrowRate: RAY.multipliedBy(3).toString(),
      availableLiquidity: '0',
      stableBorrowRate: RAY.multipliedBy(2).toString(),
      liquidityRate: RAY.multipliedBy(1).toString(),
      totalPrincipalStableDebt: '0',
      totalScaledVariableDebt: '0',
      lastUpdateTimestamp: 1,
      borrowCap: '0',
      supplyCap: '0',
      debtCeiling: '0',
      debtCeilingDecimals: 2,
      isolationModeTotalDebt: '',
      eModeLtv: 6000, // 60%
      eModeLiquidationThreshold: 7000, // 70%
      eModeLiquidationBonus: 0,
    };
  }

  addLiquidity(decimal: number | string) {
    this.reserve.availableLiquidity = new BigNumber(decimal)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.availableLiquidity)
      .toString();
    return this;
  }

  addVariableDebt(decimal: number | string) {
    this.reserve.totalScaledVariableDebt = new BigNumber(decimal)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.totalScaledVariableDebt)
      .toString();
    return this;
  }

  addStableDebt(decimal: number | string) {
    this.reserve.totalPrincipalStableDebt = new BigNumber(decimal)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.totalPrincipalStableDebt)
      .toString();
    return this;
  }
}

export class UserReserveMock {
  public userReserve: UserReserveData;
  private readonly config: { decimals: number };

  constructor(config: { decimals: number } = { decimals: 18 }) {
    this.config = config;
    const reserveMock = new ReserveMock({ decimals: config.decimals });
    this.userReserve = {
      scaledATokenBalance: '0',
      usageAsCollateralEnabledOnUser: true,
      stableBorrowRate: RAY.multipliedBy(2).toString(),
      scaledVariableDebt: '0',
      principalStableDebt: '0',
      stableBorrowLastUpdateTimestamp: 1,
      reserve: {
        ...reserveMock.reserve,
        priceInMarketReferenceCurrency: (10 ** 19).toString(), // 10
        id: '0',
        symbol: '0',
        usageAsCollateralEnabled: true,
        underlyingAsset: '0x0000000000000000000000000000000000000000',
        name: '',
        eModeCategoryId: 1,
      },
    };
  }

  supply(amount: number | string) {
    this.userReserve.scaledATokenBalance = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.scaledATokenBalance)
      .toString();
    this.userReserve.reserve.availableLiquidity = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.reserve.availableLiquidity)
      .toString();
    return this;
  }

  variableBorrow(amount: number | string) {
    this.userReserve.scaledVariableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.scaledVariableDebt)
      .toString();
    this.userReserve.reserve.totalScaledVariableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.reserve.totalScaledVariableDebt)
      .toString();
    return this;
  }

  stableBorrow(amount: number | string) {
    this.userReserve.principalStableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.principalStableDebt)
      .toString();
    this.userReserve.reserve.totalPrincipalStableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.reserve.totalPrincipalStableDebt)
      .toString();
    return this;
  }
}

export class ReserveIncentiveMock {
  public reserveIncentive: ReservesIncentiveDataHumanized;

  constructor() {
    this.reserveIncentive = {
      underlyingAsset: '0x0000000000000000000000000000000000000000',
      aIncentiveData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        rewardsTokenInformation: [
          {
            rewardTokenSymbol: 'Test',
            emissionPerSecond: (10 ** 18).toString(), // 1
            incentivesLastUpdateTimestamp: 1,
            tokenIncentivesIndex: RAY.toString(),
            emissionEndTimestamp: 1,
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            precision: 18,
            rewardPriceFeed: (10 ** 9).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0000000000000000000000000000000000000000',
          },
        ],
      },
      vIncentiveData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        rewardsTokenInformation: [
          {
            rewardTokenSymbol: 'Test',
            emissionPerSecond: (10 ** 18).toString(), // 1
            incentivesLastUpdateTimestamp: 1,
            tokenIncentivesIndex: RAY.toString(),
            emissionEndTimestamp: 1,
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            precision: 18,
            rewardPriceFeed: (10 ** 9).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0000000000000000000000000000000000000000',
          },
        ],
      },
      sIncentiveData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        rewardsTokenInformation: [
          {
            rewardTokenSymbol: 'Test',
            emissionPerSecond: '0',
            incentivesLastUpdateTimestamp: 1,
            tokenIncentivesIndex: '0',
            emissionEndTimestamp: 1,
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            precision: 18,
            rewardPriceFeed: '0',
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0000000000000000000000000000000000000000',
          },
        ],
      },
    };
  }
}

export class UserIncentiveMock {
  public userIncentive: UserReservesIncentivesDataHumanized;

  constructor() {
    this.userIncentive = {
      underlyingAsset: '0x0000000000000000000000000000000000000000',
      aTokenIncentivesUserData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        userRewardsInformation: [
          {
            tokenIncentivesUserIndex: '0',
            userUnclaimedRewards: '1',
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            rewardPriceFeed: (10 ** 19).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0',
            rewardTokenSymbol: 'Test',
          },
        ],
      },
      vTokenIncentivesUserData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        userRewardsInformation: [
          {
            tokenIncentivesUserIndex: '0',
            userUnclaimedRewards: '1',
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            rewardPriceFeed: (10 ** 19).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0',
            rewardTokenSymbol: 'Test',
          },
        ],
      },
      sTokenIncentivesUserData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        userRewardsInformation: [
          {
            tokenIncentivesUserIndex: '0',
            userUnclaimedRewards: '1',
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            rewardPriceFeed: (10 ** 19).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0',
            rewardTokenSymbol: 'Test',
          },
        ],
      },
    };
  }
}
