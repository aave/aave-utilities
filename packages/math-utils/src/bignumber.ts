import { BigNumber } from 'bignumber.js';

export type BigNumberValue = string | number | BigNumber;

export const BigNumberZeroDecimal = BigNumber.clone({
  DECIMAL_PLACES: 0,
  ROUNDING_MODE: BigNumber.ROUND_DOWN,
});

export function valueToBigNumber(amount: BigNumberValue): BigNumber {
  if (amount instanceof BigNumber) {
    return amount;
  }

  return new BigNumber(amount);
}

export function valueToZDBigNumber(amount: BigNumberValue): BigNumber {
  return new BigNumberZeroDecimal(amount);
}

export function normalize(n: BigNumberValue, decimals: number): string {
  return normalizeBN(n, decimals).toString(10);
}

export function normalizeBN(n: BigNumberValue, decimals: number): BigNumber {
  return valueToBigNumber(n).shiftedBy(decimals * -1);
}
