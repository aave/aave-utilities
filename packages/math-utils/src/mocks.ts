import BigNumber from 'bignumber.js';
import {
  ReservesIncentiveDataHumanized,
  UserReservesIncentivesDataHumanized,
} from './formatters/incentive';
import { FormatReserveUSDResponse, ReserveData } from './formatters/reserve';
import { UserReserveData } from './formatters/user';
import { RAY } from './ray.math';

export class ReserveMock {
  public reserve: ReserveData;

  private readonly config: { decimals: number };

  constructor(config: { decimals: number } = { decimals: 18 }) {
    this.config = config;
    this.reserve = {
      id: '0x0',
      symbol: 'TEST',
      name: 'TEST',
      decimals: config.decimals,
      underlyingAsset: '0x0',
      usageAsCollateralEnabled: true,
      eModeCategoryId: 1,
      reserveFactor: '0',
      baseLTVasCollateral: '5000', // 50%
      averageStableRate: '0',
      stableDebtLastUpdateTimestamp: 1,
      liquidityIndex: RAY.toString(),
      reserveLiquidationThreshold: '6000', // 60%
      reserveLiquidationBonus: '0',
      variableBorrowIndex: RAY.toString(),
      variableBorrowRate: RAY.multipliedBy(3).toString(),
      availableLiquidity: '0',
      stableBorrowRate: RAY.multipliedBy(2).toString(),
      liquidityRate: RAY.multipliedBy(1).toString(),
      totalPrincipalStableDebt: '0',
      totalScaledVariableDebt: '0',
      lastUpdateTimestamp: 1,
      borrowCap: '0',
      supplyCap: '0',
      debtCeiling: '0',
      debtCeilingDecimals: 2,
      isolationModeTotalDebt: '',
      eModeLtv: 6000, // 60%
      eModeLiquidationThreshold: 7000, // 70%
      eModeLiquidationBonus: 0,
      unbacked: '0',
    };
  }

  addLiquidity(amount: number | string) {
    this.reserve.availableLiquidity = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.availableLiquidity)
      .toString();
    return this;
  }

  addVariableDebt(amount: number | string) {
    this.reserve.totalScaledVariableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.totalScaledVariableDebt)
      .toString();
    return this;
  }

  addStableDebt(amount: number | string) {
    this.reserve.totalPrincipalStableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.totalPrincipalStableDebt)
      .toString();
    return this;
  }

  addUnbacked(amount: number | string) {
    this.reserve.unbacked = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .toString();
    return this;
  }
}

export class UserReserveMock {
  public userReserve: UserReserveData;
  public reserve: FormatReserveUSDResponse;
  private readonly config: { decimals: number };

  constructor(config: { decimals: number } = { decimals: 18 }) {
    this.config = config;
    this.userReserve = {
      underlyingAsset: '0x0000000000000000000000000000000000000000',
      scaledATokenBalance: '0',
      usageAsCollateralEnabledOnUser: true,
      stableBorrowRate: RAY.multipliedBy(2).toString(),
      scaledVariableDebt: '0',
      principalStableDebt: '0',
      stableBorrowLastUpdateTimestamp: 1,
    };
    const reserveMock = new ReserveMock({ decimals: config.decimals });
    this.reserve = {
      ...reserveMock.reserve,
      underlyingAsset: '0x0000000000000000000000000000000000000000',
      id: '0x0',
      symbol: 'TEST',
      name: 'TEST',
      decimals: config.decimals,
      usageAsCollateralEnabled: true,
      eModeCategoryId: 1,
      formattedBaseLTVasCollateral: '0.5',
      liquidityIndex: RAY.toString(),
      formattedReserveLiquidationThreshold: '0.6',
      formattedReserveLiquidationBonus: '0',
      variableBorrowIndex: RAY.toString(),
      variableBorrowRate: RAY.multipliedBy(3).toString(),
      formattedAvailableLiquidity: '0',
      liquidityRate: RAY.multipliedBy(1).toString(),
      totalPrincipalStableDebt: '0',
      totalScaledVariableDebt: '0',
      lastUpdateTimestamp: 1,
      debtCeiling: '0',
      debtCeilingDecimals: 2,
      isolationModeTotalDebt: '',
      formattedEModeLtv: '0.6',
      formattedEModeLiquidationThreshold: '0.7',
      formattedEModeLiquidationBonus: '0',
      priceInUSD: '10',
      totalLiquidityUSD: '0',
      availableLiquidityUSD: '0',
      totalDebtUSD: '0',
      totalVariableDebtUSD: '0',
      totalStableDebtUSD: '0',
      borrowCapUSD: '0',
      supplyCapUSD: '0',
      unbackedUSD: '0',
      priceInMarketReferenceCurrency: (10 ** 19).toString(),
      formattedPriceInMarketReferenceCurrency: '10',
      unborrowedLiquidity: '0',
      supplyAPY: '0',
      supplyAPR: '0',
      variableBorrowAPY: '0',
      variableBorrowAPR: '0',
      stableBorrowAPY: '0',
      stableBorrowAPR: '0',
      borrowUsageRatio: '0',
      supplyUsageRatio: '0',
      totalStableDebt: '0',
      totalVariableDebt: '0',
      totalDebt: '0',
      totalLiquidity: '0',
      debtCeilingUSD: '0',
      availableDebtCeilingUSD: '0',
      isolationModeTotalDebtUSD: '0',
      isIsolated: true,
    };
  }

  supply(amount: number | string) {
    this.userReserve.scaledATokenBalance = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.scaledATokenBalance)
      .toString();
    this.reserve.availableLiquidity = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.availableLiquidity)
      .toString();
    return this;
  }

  variableBorrow(amount: number | string) {
    this.userReserve.scaledVariableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.scaledVariableDebt)
      .toString();
    this.reserve.totalScaledVariableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.totalScaledVariableDebt)
      .toString();
    return this;
  }

  stableBorrow(amount: number | string) {
    this.userReserve.principalStableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.userReserve.principalStableDebt)
      .toString();
    this.reserve.totalPrincipalStableDebt = new BigNumber(amount)
      .shiftedBy(this.config.decimals)
      .plus(this.reserve.totalPrincipalStableDebt)
      .toString();
    return this;
  }
}

export class ReserveIncentiveMock {
  public reserveIncentive: ReservesIncentiveDataHumanized;

  constructor() {
    this.reserveIncentive = {
      underlyingAsset: '0x0000000000000000000000000000000000000000',
      aIncentiveData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        rewardsTokenInformation: [
          {
            rewardTokenSymbol: 'Test',
            emissionPerSecond: (10 ** 18).toString(), // 1
            incentivesLastUpdateTimestamp: 1,
            tokenIncentivesIndex: RAY.toString(),
            emissionEndTimestamp: 2,
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            precision: 18,
            rewardPriceFeed: (10 ** 9).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0000000000000000000000000000000000000000',
          },
        ],
      },
      vIncentiveData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        rewardsTokenInformation: [
          {
            rewardTokenSymbol: 'Test',
            emissionPerSecond: (10 ** 18).toString(), // 1
            incentivesLastUpdateTimestamp: 1,
            tokenIncentivesIndex: RAY.toString(),
            emissionEndTimestamp: 2,
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            precision: 18,
            rewardPriceFeed: (10 ** 9).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0000000000000000000000000000000000000000',
          },
        ],
      },
      sIncentiveData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        rewardsTokenInformation: [
          {
            rewardTokenSymbol: 'Test',
            emissionPerSecond: '0',
            incentivesLastUpdateTimestamp: 1,
            tokenIncentivesIndex: '0',
            emissionEndTimestamp: 2,
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            precision: 18,
            rewardPriceFeed: '0',
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0000000000000000000000000000000000000000',
          },
        ],
      },
    };
  }
}

export class UserIncentiveMock {
  public userIncentive: UserReservesIncentivesDataHumanized;

  constructor() {
    this.userIncentive = {
      underlyingAsset: '0x0000000000000000000000000000000000000000',
      aTokenIncentivesUserData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        userRewardsInformation: [
          {
            tokenIncentivesUserIndex: '0',
            userUnclaimedRewards: '1',
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            rewardPriceFeed: (10 ** 19).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0',
            rewardTokenSymbol: 'Test',
          },
        ],
      },
      vTokenIncentivesUserData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        userRewardsInformation: [
          {
            tokenIncentivesUserIndex: '0',
            userUnclaimedRewards: '1',
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            rewardPriceFeed: (10 ** 19).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0',
            rewardTokenSymbol: 'Test',
          },
        ],
      },
      sTokenIncentivesUserData: {
        tokenAddress: '0x0000000000000000000000000000000000000000',
        incentiveControllerAddress:
          '0x0000000000000000000000000000000000000000',
        userRewardsInformation: [
          {
            tokenIncentivesUserIndex: '0',
            userUnclaimedRewards: '1',
            rewardTokenAddress: '0x0000000000000000000000000000000000000000',
            rewardTokenDecimals: 18,
            rewardPriceFeed: (10 ** 19).toString(), // 10
            priceFeedDecimals: 8,
            rewardOracleAddress: '0x0',
            rewardTokenSymbol: 'Test',
          },
        ],
      },
    };
  }
}
