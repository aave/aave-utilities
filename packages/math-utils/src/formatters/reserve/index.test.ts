import BigNumber from 'bignumber.js';
import { normalize } from '../../bignumber';
import { ReserveMock } from '../../mocks';
import { formatReserve, formatReserveUSD } from './index';

describe('formatReserve', () => {
  it('should accrue over time', () => {
    const reserve = new ReserveMock().addVariableDebt(100);

    const first = formatReserve({
      reserve: reserve.reserve,
      currentTimestamp: reserve.reserve.lastUpdateTimestamp,
    });

    const second = formatReserve({
      reserve: reserve.reserve,
      currentTimestamp:
        reserve.reserve.lastUpdateTimestamp + 60 * 60 * 24 * 350,
    });

    expect(new BigNumber(second.totalDebt).gt(first.totalDebt)).toBe(true);
  });

  it('should properly calculate utilization', () => {
    const reserve = new ReserveMock();
    // no liquidity
    const zeroLiquidity = formatReserve({
      reserve: reserve.reserve,
      currentTimestamp: reserve.reserve.lastUpdateTimestamp,
    });
    expect(zeroLiquidity.borrowUsageRatio).toEqual('0');
    expect(zeroLiquidity.supplyUsageRatio).toEqual('0');

    // no borrows
    reserve.addLiquidity(100);
    const onlyLiquidity = formatReserve({
      reserve: reserve.reserve,
      currentTimestamp: reserve.reserve.lastUpdateTimestamp,
    });
    expect(onlyLiquidity.borrowUsageRatio).toEqual('0');
    expect(onlyLiquidity.supplyUsageRatio).toEqual('0');

    // borrows
    reserve.addVariableDebt(100);
    const fiftyPercent = formatReserve({
      reserve: reserve.reserve,
      currentTimestamp: reserve.reserve.lastUpdateTimestamp,
    });
    expect(fiftyPercent.borrowUsageRatio).toEqual('0.5');
    expect(fiftyPercent.supplyUsageRatio).toEqual('0.5');

    // add unbacked supplies
    reserve.addUnbacked(200);
    const unbacked = formatReserve({
      reserve: reserve.reserve,
      currentTimestamp: reserve.reserve.lastUpdateTimestamp,
    });
    expect(unbacked.borrowUsageRatio).toEqual('0.5');
    expect(unbacked.supplyUsageRatio).toEqual('0.25');
  });

  it('should calculate usd values', () => {
    // 100 liquidity with a marketPrice of 2, should be 200
    const reserve = new ReserveMock().addLiquidity(100);

    const formattedReserve = formatReserveUSD({
      reserve: {
        ...reserve.reserve,
        priceInMarketReferenceCurrency: normalize(2, -8), // 2
      },
      currentTimestamp: reserve.reserve.lastUpdateTimestamp,
      marketReferencePriceInUsd: normalize(1, -8), // 1
      marketReferenceCurrencyDecimals: 8,
    });

    expect(formattedReserve.availableLiquidityUSD).toBe('200');
    expect(formattedReserve.priceInUSD).toBe('2');
  });
});
