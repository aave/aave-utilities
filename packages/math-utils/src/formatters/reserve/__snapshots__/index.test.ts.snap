// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatReserve DAI should return the correct response 1`] = `
Object {
  "availableLiquidity": "43133.641118657852003256",
  "baseLTVasCollateral": "0.75",
  "liquidityIndex": "1.00016444737961059057",
  "reserveFactor": "0.1",
  "reserveLiquidationBonus": "0.05",
  "reserveLiquidationThreshold": "0.8",
  "stableBorrowAPR": "0.10928437169401419784",
  "stableBorrowAPY": "0.11547951583122878182",
  "supplyAPR": "0.02677620073531209306",
  "supplyAPY": "0.02713790431950061223",
  "totalDebt": "145531.711181654053680146",
  "totalLiquidity": "188665.352300311905683402",
  "totalPrincipalStableDebt": "1",
  "totalScaledVariableDebt": "145496.831599325217573288",
  "totalStableDebt": "1",
  "totalVariableDebt": "145530.711181654053680146",
  "unborrowedLiquidity": "43133.641118657852003256",
  "utilizationRate": "0.77137486776056791364",
  "variableBorrowAPR": "0.03856874338802839568",
  "variableBorrowAPY": "0.03932217240517564451",
  "variableBorrowIndex": "1.00023285443371120965",
}
`;

exports[`formatReserve WMATIC should return the correct response 1`] = `
Object {
  "availableLiquidity": "150629528.254290021063240208",
  "baseLTVasCollateral": "0.5",
  "liquidityIndex": "1.00920114827430421106",
  "reserveFactor": "0.2",
  "reserveLiquidationBonus": "0.1",
  "reserveLiquidationThreshold": "0.65",
  "stableBorrowAPR": "0.09773341053008235974",
  "stableBorrowAPY": "0.10266878586756504545",
  "supplyAPR": "0.00574176577170011131",
  "supplyAPY": "0.00575828130257027055",
  "totalDebt": "30186360.792775159242526245",
  "totalLiquidity": "180815889.047065180305766453",
  "totalPrincipalStableDebt": "0",
  "totalScaledVariableDebt": "40102377.650818088556713088",
  "totalStableDebt": "0",
  "totalVariableDebt": "30186360.792775159242526245",
  "unborrowedLiquidity": "150629528.254290021063240208",
  "utilizationRate": "0.16694528866828649738",
  "variableBorrowAPR": "0.03341338737105765182",
  "variableBorrowAPY": "0.03397788428671542517",
  "variableBorrowIndex": "1.02739968325035004938",
}
`;

exports[`formatReserve should calculate usd values 1`] = `
Object {
  "availableLiquidity": "43133.641118657852003256",
  "availableLiquidityUSD": "40649.82653321450505835335",
  "baseLTVasCollateral": "0.75",
  "borrowCapUSD": "0.0874394147481",
  "liquidityIndex": "1.00016444737961059057",
  "reserveFactor": "0.1",
  "reserveLiquidationBonus": "0.05",
  "reserveLiquidationThreshold": "0.8",
  "stableBorrowAPR": "0.10928437169401419784",
  "stableBorrowAPY": "0.11547951583122878182",
  "supplyAPR": "0.02677620073531209306",
  "supplyAPY": "0.02713790431950061223",
  "supplyCapUSD": "0.1457323579135",
  "totalDebt": "0",
  "totalDebtUSD": "0",
  "totalLiquidity": "43133.641118657852003256",
  "totalLiquidityUSD": "40649.82653321450505835335",
  "totalPrincipalStableDebt": "1",
  "totalScaledVariableDebt": "145496.831599325217573288",
  "totalStableDebt": "0",
  "totalStableDebtUSD": "0",
  "totalVariableDebt": "0",
  "totalVariableDebtUSD": "0",
  "unborrowedLiquidity": "43133.641118657852003256",
  "utilizationRate": "0",
  "variableBorrowAPR": "0.03856874338802839568",
  "variableBorrowAPY": "0.03932217240517564451",
  "variableBorrowIndex": "1.00023285443371120965",
}
`;
